### PR TITLE
Emit warning when variable is being expanded to function call

### LINKED ISSUE #3268: Warn if variable is used as function call
Today, Elixir will transform `var` into `var()` in case `var` does not exist in the user context. I would like to propose those cases to emit a warning from today on.

The downsides is that we add a tiny inconsistency. We can make function call without parens unless it is a function with null-arity. On the other hand, we already have inconsistency in the sense we can only shadow function calls with null-arity by variables. So we are exchanging one by the other.

I have bitten by this issue twice today and I believe a warning to be a far compromise.

### PR DESCRIPTION
**Pull request by lexmag**
Closes #3268.

### CODE REVIEW

**Comment by josevalim** [comment]
I think the goal is to warn only if inside a function, i.e. `E#elixir_env.function` is not nil.

File: lib/elixir/src/elixir_exp.erl [hunk 1]
```diff
@@ -313,6 +313,8 @@ expand({Name, Meta, Kind} = Var, #{vars := Vars} = E) when is_atom(Name), is_ato
           compile_error(Meta, ?m(E, file), "expected var \"~ts\"~ts to expand to an existing variable "
                         "or be part of a match", [Name, elixir_scope:context_info(Kind)]);
         true ->
+          Message = io_lib:format("variable ~ts is used as function call", [Name]),
```
Comment (josevalim) [inline lib/elixir/src/elixir_exp.erl]: We need to add to the error message: ", please use parentheses to remove the ambiguity and solve this warning"

**Comment by josevalim** [comment]
@lexmag just a heads up: let's move this to 1.3. There are already quite some changes in 1.1 and 1.2 is meant to be completely compatible. So 1.3 is when we can start pushing this.

**Comment by lexmag** [comment]
I'm still wondering if it'll be considered inconsistent if we emit this warning only inside functions.

**Comment by josevalim** [comment]
What do you mean by only inside functions? We should always emit it,
regardless of __ENV__.function is nil or not, no?

On Saturday, June 11, 2016, Aleksei Magusev <notifications@github.com>
wrote:

> I'm still wondering if it'll be considered inconsistent if we emit this
> warning only inside functions.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/3517#issuecomment-225371697>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbvX3_Ws-gdaBwO4IiT_WGlM7igSiks5qKtwQgaJpZM4FcX_5>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by lexmag** [comment]
Yes, but [previously (Jul 21, 2015)](https://github.com/elixir-lang/elixir/pull/3517#issuecomment-123188046) we thought "to warn only if inside a function".

The only downside of not checking if "function is nil" is the following:
screen shot 2016-06-11 at 18 32 44

**Comment by josevalim** [comment]
:shipit:

**Comment by tuvistavie** [comment]
When compiling a project using the `master` branch, I found out this warning, which I think is good in almost all cases.
However, I am a little concerned about DSLs.
For example, to define an Ecto schema, this forces to write:

```elixir
  schema "my_schema" do
    field :my_field, :string

    timestamps()
  end
```

which I think does not look very DSL like.
So basically I think this change would remove the syntax-sugar for all booleans attributes in DSLs.

I did not look at the implementation details yet, but do you think it would be possible and worth to handle such a case?

**Comment by whatyouhide** [comment]
@tuvistavie `timestamps` in the DLS you showed presents the same problem that this warning is trying to solve: if you have `timestamps = :foo` a few lines up, it will override your `timestamp` function (macro) call. Personally I don't think it's worth to handle such a case (provided it *is* possible).

**Comment by tuvistavie** [comment]
> timestamps in the DLS you showed presents the same problem

I understand you point, and the point of this warning,
but often a DSL is used as something declarative,
where overriding something by mistake is very unlikely to happen.
For example, let's say we have a DSL to write HTML.

```elixir
html do
  body do
    h1 "foobar"
    br()
  end
end
```

Having to make an explicit function call here kind of looses the beauty of the DSL in my opinion.

**Comment by josevalim** [comment]
@tuvistavie it depends on the DSL. Some DSLs will bypass this warning by simply annotating those variables as generated or by translating the AST. For example, `foo |> bar |> baz` won't warn because `bar` is never really called as is.

**Comment by tuvistavie** [comment]
@josevalim Thank you for the explanation.
So I suppose the best thing to do is to handle this at the DSL level
by changing these kind of calls into macros and making sure we don't trigger the warning, right?

**Comment by josevalim** [comment]
Yes, that's a possibility, definitely. We will continue collecting feedback on this in any case. :)

**Comment by ivanvotti** [comment]
I use Elixir 1.3.2 and I don't see the warning when I try this code:
```elixir
defmodule WarningTest do
  defp do_stuff, do: 'stuff'

  def my_func do
    do_stuff
    a = do_stuff
  end
end
```

**Comment by lexmag** [comment]
@ivanvotti this warning will be part of Elixir v1.4.0.

**Comment by hellerve** [comment]
Late to the party, but I cannot seem to find this in the v1.4.0 changelog. Is there any reason this was omitted?

**Comment by josevalim** [comment]
> [Kernel] Warn if variable is used as a function call

https://github.com/elixir-lang/elixir/blob/v1.4/CHANGELOG.md

**Comment by hellerve** [comment]
Okay, searched for bare words and arity. Sorry for the noise, then.

**Comment by josevalim** [comment]
No problem!

**Pull request merged**